### PR TITLE
Find kth largest in a list

### PR DESCRIPTION
**Pull request by ahdbasan**
---
Ahd AbdelRahim: Find kth largest in a list
About: find the kth largest element in a list of strings
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Adamx090 — CHANGES_REQUESTED** [review]

File: solutions/kth_largest.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    '4'
+
+    >>> kth_largest(["3", "2", "3", "1", "2", "4", "5", "5", "6"], 9)
+    '1'
+
+    >>> kth_largest([], 1)
+    Traceback (most recent call last):
+    IndexError: k must be between 1 and the length of the list
+    """
+
```
[the remark this answers is not in the document]
Comment (ahdbasan) [reply solutions/kth_largest.py:42]: Can I also say?:
```
assert isinstance(nums, list) and len(nums) > 0  # Must be non-empty list
assert all(num.isdigit() for num in nums)        # All must be digits only
assert all(num[0] != "0" for num in nums)        # No leading zeros
assert isinstance(k, int)         # k must be integer
assert k > 0                  # k must be positive
```

[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/kth_largest.py]: Trust me, this challenge is full of hidden tricks. I was overcomplicating things at first, but then I kept asking myself if there was a simpler way, and every time, I found one. So lets go through your question and check things one by one:

- Can you use hashtag comments for assertion messages? Answer is no, it must be a string. If you do that, python will treated it as a regular comment and won't display the assertion message. Assertion format must be as follows: `assert statement, "Assertion Message"`. The comma is also necessary.
- Can you test two statements with one assertion using `and`? It's possible, but avoid doing that, because you only have one assertion message, and you won't know which statement failed. Each assertion must test a single assumption, and it's okay if you have a long list of assertion, this will only help you.
- Can you use the format  `assert all(statement for item in list)` instead of the regular `for_in_:` block? Yes, actually this is shorter, I will start using it from now on. Thanks, I am still not that familiar with all these possible formats.
- Can you remove `.lstrip("+-")` from the assertion statement? No, it's important to assert that strings with negative integers are still integers, because `.isdigit()` alone only checks if all characters in a string are digits (0-9), and it returns True if only digits are found, and False if there is anything else including - and + signs. If your list has let's say, `"-5"`,  `"-5".isdigit()` will be False. That's where  `.lstrip("+-")` comes in, it simply removes any leading + and - signs and sends the the string to  `.isdigit()` with digits only, so `"-5".lstrip("+-").isdigit()` will return True.
-  `assert all(num[0] != "0" for num in nums)`. Sorry I neglected that the challenge specifies no leading zeros. Leading zeros won't impact the function behavior, but if it is specified in the challenge, it must be included. Remember, the review checklist above specifies that each assumption about an argument must have one assertion. Which reminds me, we have not asserted that the list items must be strings: `assert all(isinstance(num, str) for num in nums)`. Can you believe this? I told you this challenge is full of tricks. Let's state the list of assumptions about the arguments to summarize:

   - First argument must a be list, must be checked with an assertion.
   - The list can't be empty, must be checked with an assertion.
   - The list items must be strings, must be checked with an assertion.
   - The strings must consist of integers only, must be checked with an assertion.
   - The integers can't have leading zeros, must be checked with an assertion.
   - Second argument must be an integer, must be checked with an assertion.
   - The integer must be greater than zero and less than or equal to list length, must be checked with an assertion.
- So now we have seven assertions in total, with distinct messages:
```
assert isinstance(nums, list), "First argument must be a list"
assert len(nums) > 0, "List can't be empty"
assert all(isinstance(num, str) for num in nums), "List items must be strings"
assert all(num.lstrip("+-").isdigit() for num in nums), "Strings must consists of integers only"
assert all(num[0] != "0" for num in nums), "Integers can't have leading zeros"
assert isinstance(k, int), "Second argument must be an integer"
assert 1 <= k <= len(nums), "k must be in the range (1 <= k <= list length)"
```
   Each of these assertions will have a defensive unittest. OMG! That's a lot. But I am learning a lot also, this challenge is fun.

[the remark this answers is not in the document]
Comment (ahdbasan) [reply solutions/kth_largest.py]: I had this thought that a lot of assertions will only complicate the function that's why, **but it turns out quite the opposite; especially if we have a lot of constraints like in this challenge; It only made it clearer and simpler**

_Your review is very clear; I always had a hard time understanding assertions that's why I chose this challenge, I have learned alot from you, thankyou_

[the remark this answers is not in the document]
Comment (ahdbasan) [reply solutions/kth_largest.py]: + Question: Also, why not specify the type of error? Will that complicate the function?

[the remark this answers is not in the document]
Comment (ahdbasan) [reply solutions/kth_largest.py]: **Now, I am adding a defensive unittest for each assertion, Amazing what unittests can hold!!**

[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/kth_largest.py]: You mean use `raise` instead of `assert`, and then raise the specific type of error? You can do that. `assert` implementation is simpler though, you can use it without an `if` clause, and it's easier for debugging during development. But I can't think of anything `assert` does that `raise` can't.

[the remark this answers is not in the document]
Comment (Adamx090) [reply solutions/kth_largest.py]: The code looks great now, you just need to fix this last CI checks before we can merge.

**Review by Adamx090 — APPROVED** [review]
